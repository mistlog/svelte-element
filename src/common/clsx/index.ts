/**
 * from https://github.com/lukeed/clsx
 */
function toVal(mix) {
	var k, y, str = '';

	if (typeof mix === 'string' || typeof mix === 'number') {
		str += mix;
	} else if (typeof mix === 'object') {
		if (Array.isArray(mix)) {
			for (k = 0; k < mix.length; k++) {
				if (mix[k]) {
					if (y = toVal(mix[k])) {
						str && (str += ' ');
						str += y;
					}
				}
			}
		} else {
			for (k in mix) {
				if (mix[k]) {
					str && (str += ' ');
					str += k;
				}
			}
		}
	}

	return str;
}

export type ClassValue = ClassArray | ClassDictionary | string | number | null | boolean | undefined;

export interface ClassDictionary {
	[id: string]: any;
}

export interface ClassArray extends Array<ClassValue> { }

export function Î£(...args: ClassArray) {
	var i = 0, tmp, x, str = '';
	while (i < arguments.length) {
		if (tmp = arguments[i++]) {
			if (x = toVal(tmp)) {
				str && (str += ' ');
				str += x
			}
		}
	}
	return str;
}
