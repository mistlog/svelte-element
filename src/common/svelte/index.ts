import { setContext as rawSetContext, getContext as rawGetContext } from "svelte";

export function setContext<T>(key: keyof T, value: T[keyof T]) {
    return rawSetContext(key, value);
}

export function getContext<T>(key: keyof T): T[keyof T] {
    return rawGetContext(key);
}

export type SvelteInternalProps = { $$slots?: any };
