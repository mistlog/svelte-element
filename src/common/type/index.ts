import { Readable, Writable } from "svelte/store";

export type InputOnChangeHandler = (event: Svelte.ChangeEvent<HTMLInputElement>) => void;
export type ButtonOnClickHandler = (event: Svelte.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

type TypeWithGeneric<T> = Readable<T> | Writable<T>;
export type AutoSubscribe<Type> = Type extends TypeWithGeneric<infer X> ? X : never;

export type IconName = "edit" | "share" | "delete" | "search" | "arrow-left";