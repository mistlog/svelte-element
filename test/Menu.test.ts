import { render, fireEvent } from "@testing-library/svelte";
import { normalize } from "./common/utility";

import HorizontalMenu from "./assets/HorizontalMenu.svelte";
import VerticalMenu from "./assets/VerticalMenu.svelte";

describe("Menu", () => {
    test("horizontal menu", () => {
        const { container } = render(HorizontalMenu);
        const menu = container.querySelector(".el-menu");
        expect(normalize(menu.outerHTML)).toMatchSnapshot();
    });

    test("select menu item", async () => {
        const { container } = render(HorizontalMenu);

        const [tab1, tab2] = container.querySelectorAll(".el-menu-item");
        await fireEvent(tab2, new MouseEvent("click", {}));

        const menu = container.querySelector(".el-menu");
        expect(normalize(menu.outerHTML)).toMatchSnapshot();
    });

    test("vertical menu", () => {
        const { container } = render(VerticalMenu);
        const menu = container.querySelector(".el-menu");
        expect(normalize(menu.outerHTML)).toMatchSnapshot();
    });

    test("open submenu", async () => {
        const { container } = render(VerticalMenu);

        const submenuTitle = container.querySelector(".el-submenu__title");
        await fireEvent(submenuTitle, new MouseEvent("click", {}));

        const menu = container.querySelector(".el-menu");
        expect(normalize(menu.outerHTML)).toMatchSnapshot();
    });
});
