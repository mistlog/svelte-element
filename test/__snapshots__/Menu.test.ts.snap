// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu horizontal menu 1`] = `
"<ul class=\\"el-menu el-menu--horizontal\\" style=\\"border-bottom: none\\">
  <li class=\\"el-menu-item is-active\\">Tab1</li>
  <li class=\\"el-menu-item\\">Tab2</li>
</ul>
"
`;

exports[`Menu open submenu 1`] = `
"<ul class=\\"el-menu el-menu--vertical\\" style=\\"border-bottom: none\\">
  <li class=\\"el-submenu is-opened\\">
    <div class=\\"el-submenu__title\\">
      Group A
      <i class=\\"el-submenu__icon-arrow el-icon-arrow-down\\"></i>
    </div>
    <div style=\\"animation: __svelte_276924520_0 400ms linear 0ms 1 both\\">
      <ul class=\\"el-menu el-menu--inline\\">
        <li class=\\"el-menu-item is-active\\">Tab1</li>
        <li class=\\"el-menu-item\\">Tab2</li>
      </ul>
    </div>
  </li>
</ul>
"
`;

exports[`Menu select menu item 1`] = `
"<ul class=\\"el-menu el-menu--horizontal\\" style=\\"border-bottom: none\\">
  <li class=\\"el-menu-item\\">Tab1</li>
  <li class=\\"el-menu-item is-active\\">Tab2</li>
</ul>
"
`;

exports[`Menu vertical menu 1`] = `
"<ul class=\\"el-menu el-menu--vertical\\" style=\\"border-bottom: none\\">
  <li class=\\"el-submenu\\">
    <div class=\\"el-submenu__title\\">
      Group A
      <i class=\\"el-submenu__icon-arrow el-icon-arrow-down\\"></i>
    </div>
  </li>
</ul>
"
`;
